src/index.ts

```diff
@@ -53,15 +54,19 @@ export class Iterable<T> implements LibIterable<T> {
 
         const set = new Set<T>();
         const _that = this;
-        return new Iterable<T>((function* distinctGenerator() {
-            for (const item of _that._source) {
-                const key = keySelector(item);
-                if (!set.has(key)) {
-                    set.add(key);
-                    yield item;
-                }
+        return new Iterable<T>({
+            [Symbol.iterator]: () => {
```

Comment:
Can you change the Iterable constructor to take in a `() => Iterator<T>`  as well? That'd make Iterable simpler to consume (`new Iterable(function* {...})`) - there's a bunch of places in ux-renderer where we could then use Iterable to simplify code around generators.

Reply:
This would make this class faster and simpler as well - instead of dynamically creating functions, you can just define the filter/map etc. logic in a normal `private function*` and return `new Iterator(this._map.bind(this))`

Reply:
actually, why do you even need to store the surrounding Iterable object? all you need is the () => Iterator, right?

Reply:
oh crap, didn't see your last commit. let me know if this PR works for you: https://github.com/riteshrao/jsiterable/pull/3/files, I also added tests